funcx_sdk/funcx/serialize/concretes.py

```diff
@@ -124,16 +129,93 @@ class DillCode(SerializeBase):
     def __init__(self):
         super().__init__()
 
-    def serialize(self, data):
+    def serialize(self, data) -> str:
         x = codecs.encode(dill.dumps(data), "base64").decode()
         return self.identifier + x
 
-    def deserialize(self, payload):
+    def deserialize(self, payload: str):
         chomped = self.chomp(payload)
         function = dill.loads(codecs.decode(chomped.encode(), "base64"))
         return function
 
 
+class CombinedCode(SerializeBase):
+    """This method uses multiple methods to serialize a function
+
+    # These should be a superset of the methods
+    Code from interpreter/main        : Yes
+    Code from notebooks               : Yes
+    Works with mismatching py versions: Yes
+    Decorated fns                     : Yes
+    """
+
+    identifier = "10\n"
+
+    def get_multiple_payloads(self, payload: str) -> tuple:
+        parts = self.chomp(payload).split(COMBINED_SEPARATOR)
+        assert len(parts) == 2 * len(DO_SERIALIZE_METHODS)
+        payloads: list[tuple[str, str]] = []
+        for index in range(len(parts) // 2):
+            payloads.append((parts[index], parts[index + 1]))
+
+        # ie. ((04\n', 'gACQ...'), ('01\n', 'sAAbk2...'), ...)
+        return tuple(payloads)
+
+    def serialize(self, data) -> str:
+        chunks = []
+        for id in DO_SERIALIZE_METHODS:
+            method = METHODS_MAP_CODE[id]()
+            serialized = method.serialize(data)
+            id_length = len(id)
+            chunks.append(serialized[:id_length])
+            chunks.append(serialized[id_length:])
+        return self.identifier + COMBINED_SEPARATOR.join(chunks)
+
+    def deserialize(self, payload: str, variation: int = 1):
+        """
+        Return the other variants, if encoding contains them,
+        if requested.  varation is 1-indexed (ie. 2 means 2nd method)
+        Currently defaults to the first variation's deserialization
+        """
+        count = 0
+        for serial_id, encoded_func in self.get_multiple_payloads(payload):
+            count += 1
+            if count == variation:
```

Comment:
This doesn't seem like a very efficient approach to list/tuple addressing?  Maybe I'm missing something obvious, but why not just directly address the tuple rather than looping through it and testing the index against variation?

E.g.:
```
serial_id, encoded_func =  self.get_multiple_payloads(payload)[variation - 1]
if serial_id in METHODS_MAP_CODE:
....
```

Reply:
After thinking about the intent here, I'm wondering if we really want to be decoding a specific variation at all.  As I understood the intent of the design here, we would try each variation until one succeeded.  The logic here (and even with the change suggested) will only ever try one variant.  I think we should drop the `variation` argument from the signature and change the try except block below to only raise an exception if we've failed to decode all variants.

Reply:
@joshbryan-globus I did originally do 'decode variants until one succeeds' but added the variant param because testing revealed that it's more common (whether or not IRL use) to have deserialization succeeding but have run() fail later.  ie. in that case, the first variation the loop encounters will succeed, but will fail later during run.  (Category 3 failures).  If the user knew which method actually deserializes AND runs for his use case, then specifying it via a param  seems to be the only solution.  Note that currently there is no such param in FuncXClient.register_function(), as I wasn't sure if we want to add the API param right now before someone complains our new (old) default DillCodeSource does not work.

Reply:
> This doesn't seem like a very efficient approach to list/tuple addressing? Maybe I'm missing something obvious, but why not just directly address the tuple rather than looping through it and testing the index against variation?
> 
> E.g.:
> 
> ```
> serial_id, encoded_func =  self.get_multiple_payloads(payload)[variation - 1]
> if serial_id in METHODS_MAP_CODE:
> ....
> ```

Yes the code could be simplified a bit, will do so if we go forward with the variation logic.

Reply:
The code was simplified (the loop code)